src/liballoc/raw_vec.rs

```diff
@@ -517,12 +578,21 @@ impl<T, A: Alloc> RawVec<T, A> {
                 return;
             }
 
-            let new_cap = self.amortized_new_size(used_cap, needed_extra_cap);
+            let elem_size = mem::size_of::<T>();
+            let new_cap = amortized_new_capacity(elem_size,
+                                                 used_cap,
+                                                 needed_extra_cap);
 
             let new_layout = match Layout::array::<T>(new_cap) {
                 Some(layout) => layout,
                 None => panic!("capacity overflow"),
             };
+            let (_, usable_size) = self.a.usable_size(&new_layout);
```

Comment:
Do we have any idea how expensive this is for the usual allocators (jemalloc, libc)?

Reply:
When one makes a memory allocation, the first thing the allocator does is determine which memory "list" it is going to put the allocation into (allocators typically have different internal lists/pools for differently-sized allocations). 

Typically, this is just a match on the allocation size and returns a constant (so should be pretty quick). The [`Alloc::usable_size`](https://doc.rust-lang.org/beta/alloc/allocator/trait.Alloc.html#method.usable_size) docs do not guarantee any time complexity for this method, but I expect it to always be `O(1)` in practice.

Because of this we might miss using some extra memory, so I wonder whether I should use [`alloc_excess`](https://doc.rust-lang.org/beta/alloc/allocator/trait.Alloc.html#method.alloc_excess) (and `realloc_excess`) instead of `usable_size` when I can, @bluss ?

Reply:
I think `realloc_excess` is right thing, I guess it's there to optimize this exact problem (or fallback gracefully). 

Although, the jemalloc implementation needs to do it properly https://github.com/rust-lang/rust/blob/master/src/liballoc_jemalloc/lib.rs#L222

Reply:
Yes, I wanted to use `nallocx` and since I only found it for `usable_size` I used that instead.

I think whether we use `realloc_excess` or `usable_size` won't affect the benchmarks for `jemalloc` since `realloc_excess` will be implemented as `nallocx + rallocx`, but we wan't to use `realloc_excess` to be more "allocator independent".

I don't expect it to be a huge difference in memory requirements between both, but there might be a performance difference.

Reply:
Indeed. There's probably some interesting cycles to be saved there (1 call frame, a couple of branches, ...) even if internally it calls 2 jemalloc apis.